lib/pbench/server/hierarchy.py

```diff
@@ -228,118 +207,124 @@ def __init__(self, name, path, config):
+                    cnt = self.dump_check(
+                        fp, controller, key, cnt, self.lead_tab, self.name
+                    )
                 if controller not in self.tarballs:
                     fp.write(
-                        "\t* No tar ball files found in this controller directory.\n"
+                        f"{self.lead_tab[:-2]}* No tar ball files found in this controller directory.\n"
                     )
-                    cnt += 1
+                    cnt = 1
         return cnt
 
-    def check_controller(self):
-        for controller in sorted(self.controllers):
+    def check_controller(self) -> bool:
+        for controller in self.controllers:
             if (
-                self.check_controller_in_list(controller)
+                self.is_controller_in_list(controller)
                 or controller not in self.tarballs
             ):
                 return True
         return False
 
-    def output_bad_controllers(self, fp):
+    def output_bad_controllers(self, fp: IO):
         fp.write("\nBad Controllers:\n")
         for controller in sorted(self.bad_controllers):
-            """Formatting output into ls -l format to align
-            output with gold files
-            """
+            """Formatting output into ls -l format"""
             contStatOb = os.stat(controller)
-            fperm = filepermissions(contStatOb.st_mode)
+            fperm = stat.filemode(contStatOb.st_mode)
             mTime = time.strftime(
                 "%a %b %d %H:%M:%S.0000000000 %Y", time.gmtime(contStatOb.st_mtime)
             )
-            fp.write(f"\t-{fperm}          0 {mTime} {controller.name}\n")
-        return
+            fp.write(f"\t{fperm}          0 {mTime} {controller.name}\n")
 
 
 class ControllerHierarchy(Hierarchy):
-    def __init__(self, name, path, config):
+    def __init__(self, name: str, path: Path, config: PbenchServerConfig):
         super().__init__(name, path, config)
 
-        self.verifylist = list()
+        self.verifylist = set()
+        self.lead_tab = "\t"
         self.validation_list = {
-            self.MIALIST: List(
+            self.MIALIST: ListOfController(
                 f"Controllers which do not have a {self.config.ARCHIVE} directory:"
             ),
-            self.EMPTY_CONTROLLERS: List("Controllers which are empty:"),
-            self.UNEXPECTED_CONTROLLERS: List(
+            self.EMPTY_CONTROLLERS: ListOfController("Controllers which are empty:"),
+            self.UNEXPECTED_CONTROLLERS: ListOfController(
                 "Controllers which have unexpected objects:"
             ),
         }
 
-    def add_controller_list(self, val, controller):
-        """Gather values for each controller_check dict key"""
-        self.validation_list[val].additem(controller)
+    def add_controller_list(self, val: str, controller: str):
+        """Gather controllers for each discrete values"""
+        self.validation_list[val].append(controller)
```

Comment:
It's kind of strange and disturbing to have the values in `validation_list` be `DictOfList`'s sometimes and `ListOfController`'s at other times.

(I don't have the strength to investigate this properly, today; I'll try again tomorrow or Friday.)